### ISSUE TITLE
KeysetScrollSpecification null properties support

### ISSUE THREAD

**Issue by ivan-zaitsev** [spring-projects/spring-data-jpa#3605]
It is documented [here](https://docs.spring.io/spring-data/keyvalue/reference/data-commons/repositories/scrolling.html#repositories.scrolling.keyset) that keyset-filtering requires the keyset properties (those used for sorting) to be non-nullable.

But will it actually affect indexing by adding `cb.isNull()` conditions to https://github.com/spring-projects/spring-data-jpa/blob/main/spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/KeysetScrollSpecification.java#L120 ?

**Comment by mp911de** [comment]
Not quite sure what you're asking for. The general issue is that various databases may have support for null precedence in ordering while others don't support null precedence. Also, JPA doesn't support the specification of null precedence. This is scheduled for the next JPA spec release.

It would help if you come up with a bit more of an example, how you envision things should work so that we can find out whether this ticket can become actionable.

**Comment by ivan-zaitsev** [comment]
@mp911de 

I am talking about `WHERE` condition for keyset-filtering. Currently it is generated without `IS NULL` conditions.

WHERE (**created_at IS NOT NULL AND** created_at > ?) OR
      (**created_at IS NULL AND** id > ?)

**Comment by schauder** [comment]
If we support `NULL` values we need to support `NULLS FIRST/LAST` semantics.

This could actually be achieved with adding an additional virtual field to the where clause generation which is `0` for non-null values and `1` or `-1` for null values. 
But that is probably really bad for the resulting explain plan. 
I assume the same can be achieved using different operators (`<, <=, >=, >`). And your query might actually be one of those solutions.

Still I'm afraid this might be bad for performance, so I think we should find a way to only apply this to fields which are nullable.

IF we get all this working, I don't think we need a "fallback" we only need the combinations of all fields in the key set to be unique.

**Comment by ivan-zaitsev** [comment]
> If we support NULL values we need to support NULLS FIRST/LAST semantics.

Of course. It is another issue which is related to ORDER BY clause which comes after WHERE conditions, because WHERE conditions alone would not solve issue completely. I believe there is another task for ORDER BY in github issues.

> Still I'm afraid this might be bad for performance, so I think we should find a way to only apply this to fields which are nullable.

> IF we get all this working, I don't think we need a "fallback" we only need the combinations of all fields in the key set to be unique.

Yes, agree. The implementation code should be in control to make field nullable not not, maybe only by default all fields will not be nullabe.

**Comment by schauder** [comment]
> Of course. It is another issue which is related to ORDER BY clause which comes after WHERE conditions, because WHERE conditions alone would not solve issue completely. I believe there is another task for ORDER BY in github issues.

The where clause is actually all about ordering. After all, we are trying to select the next `n` elements after the _current_ one. So we need to consider if `null` comes before the first `null` elements, or after in the where clause.

**Comment by quaff** [comment]
```java
Map<String, Object> keys = new LinkedHashMap<>();
keys.put("createdAt", entity.getCreatedAt());
keys.put("id", entity.getId());
```

@ivan-zaitsev FYI, `id` property is not required since Spring Data JPA will add it implicitly to make sure uniqueness, also I'm proposing to not add `id` if provided properties combination are unique, see #3013.
